src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java

```diff
@@ -134,26 +138,24 @@ public Dashboard remoteCreate(DashboardRemoteRequest request, boolean isUpdate)
 
         List<DashboardRemoteRequest.Entry> entries = request.getAllEntries();
         Map<String, WidgetRequest> allWidgetRequests = generateRequestWidgetList( entries, dashboard);
+        Set<CollectorType> existingTypes = new HashSet<>();
+        Set<CollectorType> incomingTypes = new HashSet<>();
+        Component component = componentRepository.findOne(dashboard.getApplication().getComponents().get(0).getId());
+        component.getCollectorItems().forEach((item,val)->{
+            existingTypes.add(item);
+        });
+
         //adds widgets
         for (String key : allWidgetRequests.keySet()) {
             WidgetRequest widgetRequest = allWidgetRequests.get(key);
 
-            if( key.equals( "codeanalysis" ) ){
-
-                List< CollectorItem > list = new ArrayList<>();
-                Component component = componentRepository.findOne( dashboard.getApplication().getComponents().get(0).getId() );
-
-                list.addAll( component.getCollectorItems(CollectorType.Test ));
-                list.addAll( component.getCollectorItems(CollectorType.StaticSecurityScan ) );
-                list.addAll( component.getCollectorItems(CollectorType.CodeQuality ) );
-                list.addAll( component.getCollectorItems(CollectorType.LibraryPolicy ) );
-
-                List< ObjectId > collectorItemIdList = list.stream().map( CollectorItem::getId).collect(Collectors.toList() );
-                widgetRequest.getCollectorItemIds().addAll( collectorItemIdList );
-            }
+            widgetRequest.getCollectorItemIds().forEach(id->{
+                CollectorItem  c = collectorItemRepository.findOne(id);
+                Collector collector = collectorRepository.findOne(c.getCollectorId());
```

Comment:
How do we handle bad data? assume if the collector item does not exist, then whole process will be aborted. You want to reconsider the logic

Reply:
added null-safe